### PR TITLE
Add Likes and Sharing sidebar as a Gutenberg extension

### LINKED ISSUE #10199: Likes and Shares metabox design for Gutenberg and Classic
The typography, color, and open/close caret of the "Likes and Shares" metabox in Gutenberg are different from the other ones:

likes-and-shares-metabox

For reference, this is how it looks in the classic editor (it is in line with the rest of components):

likes-and-shares-metabox-classic

I know we're in a transitional moment so things like this are still in flux. Wanted to surface it somewhere given that we're going to need to support both classic and gutenberg in the foreseable future. Please, feel free to close if this is being worked / tracked in any other place.

### PR DESCRIPTION
**Pull request by codebykat**
#### Changes proposed in this Pull Request:

This PR converts the Likes and Sharing metabox to a Gutenberg sidebar extension, moving these settings out of the main sidebar and into the Jetpack section. It is mostly a port of https://github.com/Automattic/wp-calypso/pull/29744, with the addition of `__back_compat_meta_box => true` to disable 
the old metabox within Gutenberg.

It also adds some logic so that the checkboxes aren't displayed unless their corresponding modules are active (because the extension will be added if *either* Sharing or Likes is activated, but we don't want to show checkboxes for a disabled setting).

Fixes #10199

#### Testing instructions:

1. Double-check [sharing button prerequisites](https://github.com/Automattic/wp-calypso/pull/29744#issuecomment-459443781).
1. Make sure beta blocks are enabled with `define( 'JETPACK_BETA_BLOCKS', true );` in your `wp-config.php`.
1. Compile the blocks with `yarn build-extensions` in the checked-out branch.
1. Load a new or existing post in the block editor
1. The Likes and Sharing checkboxes should no longer be visible in the post sidebar under Discussion.
1. Under the Jetpack sidebar, there should be a Likes and Shares section with checkboxes to toggle these options:

screen shot 2019-01-24 at 3 44 01 pm

5. Make sure you can set/unset them and that these settings are appropriately reflected on the post (you can check via the API console or the classic editor, or by just viewing the post on the front-end).

6. Also test by disabling Likes globally and then toggling them ON on a per-post basis.

7. In the Classic Editor, verify that the checkboxes still show up and behave correctly.

8. **New! Disable one or both modules and verify that the checkboxes are shown or hidden as appropriate.**

#### Proposed changelog entry for your changes:

* Convert Likes and Sharing metabox to a Gutenberg plugin

### CODE REVIEW

**Review by roccotripaldi — DISMISSED** [review]
This works well. I think we can address your proposed follow up in an other PR

**Comment by kwight** [comment]
Once I got things built with `yarn build-extensions` and the beta blocks allowed, I was able to get the settings to show up in the JP sidebar, but the old ones still display in the original sidebar. `__back_compat_meta_box` seems ineffective.

**Comment by kwight** [comment]
> but the old ones still display in the original sidebar. __back_compat_meta_box seems ineffective.

Ignore me – I was mistaking the Discussion metabox for Likes and Sharing. All good.

**Review by kwight — DISMISSED** [review]

**Comment by codebykat** [comment]
I've updated this PR to hide irrelevant checkboxes based on which modules are active, and to follow the patterns in the VideoPress extension for determining availability and registering the sidebar extension.

**Comment by sirreal** [comment]
I pushed two small changes:
- Remove the `@format` pragmas (no longer needed)
- Categorize and sort some import correctly under internal/external headers

**Comment by jeherve** [comment]
this will now need a rebase I'm afraid.

**Review by sirreal — DISMISSED** [review]
I gave this quite a bit of testing and everything behaved as expected. It was difficult for me to get the modules enabled, I ended up having to go to the debug module view and enable them.

Once I got things set up, I tested lots of combinations and everything worked well 👍 

I did not test the legacy meta boxes.

**Review by jeherve — DISMISSED** [review]
This seems to work well. I only have one remark.

File: class.jetpack-gutenberg.php
Comment (codebykat) [reply class.jetpack-gutenberg.php]: That makes sense, I'll verbos-ify it. Thanks @jeherve

Comment (codebykat) [reply class.jetpack-gutenberg.php]: Done and ready for review again!

**Comment by codebykat** [comment]
Note that the real WPCOM patch for this is D26061-code. This includes the synced file changes as well as the built blocks and all the WPCOM-specific stuff.

File: class.jetpack-gutenberg.php
Comment (ockham) [inline class.jetpack-gutenberg.php]: We were hoping for `Jetpack_Editor_Initial_State` to be the exception rather than the rule to communicate information from the server to the client, i.e. to not really set an example for people to add more fields to :sweat_smile: Would it be possible to either

- Use a mechanism as provided by `@wordpress/data` to fetch information from the network on the client side, or
- Use our existing `available_blocks` mechanism?

My gut feeling tells me that the latter might be a good fit for this scenario: Since there already are distinct modules for both Likes and Shares, we could expose them both through separate `Jetpack_Gutenberg::set_extension_available()` calls on the server.

On the client side, we could then
- move each the Likes and Shares checkbox to a file of their own
- and use the [Slot/Fill](https://wordpress.org/gutenberg/handbook/designers-developers/developers/components/slot-fill/) technique to 
  - turn the `LikesAndSharesPanel` component into a `<Slot />` (akin to our [`< JetpackPluginSidebar />`](https://github.com/Automattic/jetpack/blob/c9dc14ac0b6af78c23400f3383e3ae442b48f7f7/extensions/shared/jetpack-plugin-sidebar.js))
  - and conditionally add each checkbox to it.

I think this could be reasonably straight-forward -- let me know if I can help you with implementing it.

Comment (codebykat) [reply class.jetpack-gutenberg.php]: Ah okay. FYI @roccotripaldi since this implementation was your suggestion -- I guess it's not recommended.

I'll see if I can get it working as you suggest.

Comment (roccotripaldi) [reply class.jetpack-gutenberg.php]: ah sorry for the bad recommendation. lesson learned.

Comment (sirreal) [reply class.jetpack-gutenberg.php]: This is a good point, sorry I didn't spot it in my earlier review.

I'll see about adding a comment to this block in another PR to clarify why it exists and that is discouraged as a general purpose mechanism.

**Review by jeherve — DISMISSED** [review]
Things seem to work well for me now. 👍

File: class.jetpack-gutenberg.php
Comment (roccotripaldi) [reply class.jetpack-gutenberg.php]: I attempted to address this in https://github.com/Automattic/jetpack/pull/11709/commits/7a0df200bb1587bcf809eb4f1f4635542bd9aae0, but now I cannot seem to get anything to render. Maybe Gutenberge doesn't like nested slot/fills. Any ideas here?

Comment (roccotripaldi) [reply class.jetpack-gutenberg.php]: Never mind. I figured it out!

**Comment by roccotripaldi** [comment]
This is ready for reviews again y'all.

File: modules/likes/jetpack-likes-settings.php
Comment (mmtr) [inline modules/likes/jetpack-likes-settings.php]: Note that this is disabling the default metabox even if the Jetpack beta blocks are disabled. We should check if the `likes` extension is available before disabling the old metaboxes.

File: modules/sharedaddy/sharedaddy.php
Comment (mmtr) [inline modules/sharedaddy/sharedaddy.php]: Same here. We should check if the `sharing` extension is available before disabling it.

**Review by mmtr — COMMENTED** [review]
This looks almost perfect! Wonderful job here @codebykat and @roccotripaldi!

Tested different scenarios and they all worked as expected as long as you have enabled the Jetpack beta blocks. Left a comment about handling environments where the beta blocks are not available.

File: extensions/shared/jetpack-likes-and-sharing-panel.js
Comment (sirreal) [inline extensions/shared/jetpack-likes-and-sharing-panel.js]: The wrapper is no more: #11743

We'll need to update this and merge with external deps above:

```suggestion
import { __ } from '@wordpress/i18n';
```

File: extensions/blocks/likes/likes-checkbox.js
Comment (sirreal) [inline extensions/blocks/likes/likes-checkbox.js]: Please merge this with external deps:

```suggestion
import { __ } from '@wordpress/i18n';
```

File: extensions/blocks/sharing/sharing-checkbox.js
Comment (sirreal) [inline extensions/blocks/sharing/sharing-checkbox.js]: Merge with external deps:

```suggestion
import { __ } from '@wordpress/i18n';
```

File: extensions/shared/jetpack-likes-and-sharing-panel.js
Comment (sirreal) [inline extensions/shared/jetpack-likes-and-sharing-panel.js]: Needs text domain now:

```suggestion
				<PanelBody title={ __( 'Likes and Sharing', 'jetpack' ) }>
```

File: extensions/blocks/likes/likes-checkbox.js
Comment (sirreal) [inline extensions/blocks/likes/likes-checkbox.js]: Needs text domain now.

File: extensions/blocks/sharing/sharing-checkbox.js
Comment (sirreal) [inline extensions/blocks/sharing/sharing-checkbox.js]: Needs text domain now.

**Review by sirreal — DISMISSED** [review]
We need an update now that the i18n wrapper has been dropped 😬

File: extensions/blocks/likes/index.js
Comment (sirreal) [inline extensions/blocks/likes/index.js]: Observation (this is fine as is — no need to change): If an arrow function returns a single expression, we can use the short form:

```js
	render: () => (
		<JetpackLikesAndSharingPanel>
			<LikesCheckbox />
		</JetpackLikesAndSharingPanel>
	),
```

**Comment by simison** [comment]
Noting that we could move these files to planned `/extensions/plugins/` folder instead of `/extensions/blocks` but I'm ok leaving it here until we merge this. We anyway need to move Publicize and SEO description so it's not a big deal to leave moving this for later. There could be hiccups in the folder switch and I wouldn't want to mix those issues into this PR at this point.

File: extensions/blocks/likes/index.js
Comment (ockham) [reply extensions/blocks/likes/index.js]: Since this function is an object property, and isn't referencing `this` in the function body, we might alternatively shorten to this (method-like) notation:

```suggestion
	render() {
```

File: extensions/blocks/likes/likes-checkbox.js
Comment (ockham) [inline extensions/blocks/likes/likes-checkbox.js]: I think `onChange={ toggleLikes }` should be sufficient here :slightly_smiling_face:

File: extensions/blocks/likes/likes-checkbox.js
Comment (ockham) [inline extensions/blocks/likes/likes-checkbox.js]: Nit, could probably use a functional stateless component here

File: extensions/blocks/likes/likes-checkbox.js
Comment (ockham) [inline extensions/blocks/likes/likes-checkbox.js]: TBH I'd rather not create an ad-hoc function in `applyWithDispatch` (we even have eslint rules against that in Calypso). How about just returning `{ editPost }` and using updating its usage accordingly in the component body? (That will possibly invalidate [this comment](https://github.com/Automattic/jetpack/pull/11709/files#r271670919) :slightly_smiling_face: )

File: extensions/blocks/sharing/index.js
Comment (ockham) [inline extensions/blocks/sharing/index.js]: https://github.com/Automattic/jetpack/pull/11709/files#r271602675 / https://github.com/Automattic/jetpack/pull/11709/files#r271668386 :grimacing:

File: extensions/blocks/sharing/sharing-checkbox.js
Comment (ockham) [inline extensions/blocks/sharing/sharing-checkbox.js]: https://github.com/Automattic/jetpack/pull/11709/files#r271670919

File: extensions/blocks/sharing/sharing-checkbox.js
Comment (ockham) [inline extensions/blocks/sharing/sharing-checkbox.js]: https://github.com/Automattic/jetpack/pull/11709/files#r271671051

File: extensions/blocks/sharing/sharing-checkbox.js
Comment (ockham) [inline extensions/blocks/sharing/sharing-checkbox.js]: https://github.com/Automattic/jetpack/pull/11709/files#r271673622

File: extensions/shared/jetpack-likes-and-sharing-panel.js
Comment (ockham) [inline extensions/shared/jetpack-likes-and-sharing-panel.js]: :heart:

**Review by ockham — DISMISSED** [review]
Thanks @codebykat and @roccotripaldi, amazing work! I :heart: how clean this code is. I left a few minor suggestions, none of them blocking.

**Comment by roccotripaldi** [comment]
> Noting that we could move these files to planned /extensions/plugins/

We'll get there soon!

**Comment by roccotripaldi** [comment]
This is ready for review again y'all

**Review by obenland — APPROVED** [review]
Looks good from my end. Checkboxes show up in the Jetpack sidebar, likes & sharing buttons show up (or not) as desired. Thanks again for seeing this through @roccotripaldi!

**Comment by kwight** [comment]
Works great! Checked both Likes and Shares, individually and together, with both activated and with only one activated; all UI and front-end behaved as expected <3 Awesome job everyone!

**Review by kwight — APPROVED** [review]

**Review by mmtr — APPROVED** [review]

**Review by jeherve — APPROVED** [review]
This works well for me. I am approving, although I wonder how / when we'll be able to make the switch to Production with the new block building process? Are you planning on removing the Beta checks and moving the block from Beta to Production right before we enter the Beta period, and then synchronize that change to WordPress.com as soon as the release goes out?

**Comment by roccotripaldi** [comment]
> Are you planning on removing the Beta checks and moving the block from Beta to Production right before we enter the Beta period, and then synchronize that change to WordPress.com as soon as the release goes out?

I plan to publish Kat's call for testing, and if all goes well, I'll remove the beta check, and move this block into production.

We decided that we are not worried about keeping our WordPress.com "releases" in-line with Jetpack releases. WordPress.com code will still be continuously deployed as it has always been.

**Comment by simison** [comment]
>WordPress.com "releases" in-line with Jetpack releases.

For this specific feature it isn't so important either: it's not a block in site's content but a sidebar UI to existing feature.

**Comment by sarahfg** [comment]
There's a user reported issue in HC-2112399 for the site https://watercresswords.com/ that seems related to this. I have been unable to reproduce it on my test site, but the issue has only been present on posts published since this was merged and there are no other differences in the settings the user is using for sharing. 

The user's posts should have the sharing buttons on them, as they are checked on both the Sharing section for posts, and within the fancy new Jetpack Sharing section this added :-) . The sharing section isn't visible though on posts published since the merge. The like button is present, but not the sharing for various socials:

missing socials

(Image link: https://cld.wthms.co/0iWvVY )

To help the user I tried going in and turning sharing off through the post's Jetpack settings, updating, then turning it on again. That seemed to solve it. I wanted to share here in case there is another repeat of this issue on any other sites since this merge is the only change I could identify on the site related to sharing.

**Comment by molszewska** [comment]
Hey, we have an issue with https://faithexpressedinlove.com site.

The user can't see "Likes and Sharing" option: https://snag.gy/eU35ra.jpg in the editor. 
We asked them to clear cache, reload the editor and try another browser/computer/wifi - no luck.

We couldn't reproduce the issue, as the option appears just fine for us: https://cld.wthms.co/dmmmrs 

Could you double check if we're missing anything? Thanks!

1962477-zen

**Comment by jeherve** [comment]
@molszewska The site you are referring to is a WordPress.com site. The Likes and Sharing option is not enabled on WordPress.com sites yet. You can subscribe to D26858-code to find out when the feature will be enabled for everyone. There will most likely be an announcement as well.

**Comment by jeherve** [comment]
@mrfoxtalbot It is, but as an a11n you see things that the site owner does not (yet).

**Comment by codebykat** [comment]
FYI @molszewska and @mrfoxtalbot this feature is now live on WPCOM for everyone :)

**Pull request merged**
